src/ocean/math/WideInt.d

```
@@ -0,0 +1,474 @@
+    }
+
+    /***************************************************************************
+
+        Mutates current number in-place, dividing it by `rhs` and calculating
+        remainder.
+
+        Params:
+            rhs = number to divide by
+
+        Returns:
+            remainder of division
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
+    {
+        ulong remainder = 0;
+
+        for (ptrdiff_t idx = this.payload.length - 1; idx >= 0; --idx)
+        {
+            remainder = (remainder << 32) + this.payload[idx];
+            ulong result = remainder / rhs;
+            remainder -= rhs * result;
+            assert(result <= uint.max);
+            this.payload[idx] = cast(uint) result;
+        }
+
+        assert(remainder <= uint.max);
+        return cast(uint) remainder;
+    }
+
+    unittest
+    {
+        WideInt num = 56 * 47;
+        test!("==")(0, num.divideBy(47));
+        test!("==")(num.payload[0], 56);
+
+        num.payload[0] = 1337;
+        num.payload[1] = 100;
+        test!("==")(1337, num.divideBy(1 << 31));
+        test!("==")(num.payload[0], 200);
+    }
+
+    /***************************************************************************
+
+        Mutates this number in place multiplying it by a regular integer
+
+        Params:
+            rhs = value to multiply by
+
+        Throws:
+            WideIntRangeException if resulting value can't fit in current
+            WideInt value range
+
+    ***************************************************************************/
+
+    public void multiplyBy ( uint rhs )
+    {
+        ulong overflow = 0;
+
+        for (size_t i = 0; i < this.payload.length; ++i)
+        {
+            overflow += cast(ulong)(this.payload[i]) * rhs;
+            this.payload[i] = cast(uint) (overflow & uint.max);
+            overflow >>= 32;
+        }
+
+        enforce(.wideint_exception, overflow == 0);
```

Given that this is a wide unsigned integer, do we want this exception-on-overflow behaviour, or do we want unsigned integer-style wraparound (for consistency)?

The current behaviour looks simpler to implement, so I'm inclined to go with this personally.

---

I am quite confident we want exception in this case. Native integers have wraparound because it is how hardware works and it is fast. In almost all `WideInt` use cases you assume that it is "big enough" for all your values, essentually using it as `BigInt` with fixed size. If such assumption is violated, you want to know about it.

---

Well, hardware reports wrap-around as well, though the carry and overflow flags, but D swallows this exception.

---

No, it is not about carry (you may notice below that I have addition function that does report carry flag), it is about multiplication. When you multiply `ulong` by `ulong` your result can be up to 128bit large.

---

But that's overflow then, which is still reported by hardware but not D.

---

AFAIK hardware reports that it has happened, not overflown amount. That is sufficient for addition (where amount always == 1) but not multiplication.

---

That's right.

---

So you do agree that using `uint` for a single word is necessary? :)

---

Oh yes.